connection_pool/connection_pool.go

```diff
@@ -32,17 +32,33 @@ var (
 	ErrNoHealthyInstance = errors.New("can't find healthy instance in pool")
 )
 
+// ConnectionListener provides an observer semantics for components interested
+// in knowing the role changes of connections in a ConnectionPool.
+type ConnectionListener interface {
+	// Add is called before a connection with a role added into a list of
+	// active connections. A ConnectionPool object will close connection
```

Comment:
No, it will not. My logic was:

If `Add` returns an error, a connection will not be added to a pool -> nothing to remove.

Reply:
I think you should also clarify it in doc since it may be confusing.

Also I'm not sure that we shouldn't call `Removed` in this case. User may do something in `Add` and place all cleanup-related code in `Removed`. But, as far as I understand, user still could do it in current implementation if needed, so it's up to you.

Reply:
Let's make it symmetrical. I've update the logic: it calls `Closed` on every `Opened` at now.